### ISSUE TITLE
About sampling poses

### ISSUE THREAD

**Issue by sseunghyuns** [Ending2015a/liepose-diffusion#2]
Dear author, 

First of all, thanks for your amazing work. I have some questions regarding inference phase and SE(3) diffusion. 

1. In the inference phase, do you only sample one pose?  Given the nature of generative models, it seems possible that the sampled pose might deviate from the distribution(regarded as outlier). Have you encountered issues with this?  Or has the model been trained well enough that such concerns are unnecessary? If I missed something, please let me know!

2. And as proposed in [1], since using a 6D representation is continuous in Euclidean space, it seems we can apply the traditional diffusion training directly (Here, we are representing SO(3) with 6D-representation). Would there be any significant differences when training the model this way? Would it be better in terms of performance to train the model in a more complex SE(3)? I would like to hear your opinion!

Thank you.

[1] Yi Zhou, Connelly Barnes, Jingwan Lu, Jimei Yang, and Hao Li. On the continuity of rotation representations in neural networks. In Proceedings of the IEEE/CVF Conference on Computer Vision and Pattern Recognition, pages 5745–5753, 2019.

**Comment by Ending2015a** [comment]
Hello @sseunghyuns,
thank you for the question.

1. For evaluation, we sample 1 pose for each test data. For visualization, we samples 1000 poses from our diffusion models for each plot without using any filtering method to remove outliers. As you can see from our visualization, there are almost no outliers. Outliers usually occur when the model is not well-trained (or well-tuned, e.g. noise scheduling, hyperparameters) or when the dataset lacks sufficient data to cover the whole SO(3) and SE(3). In the latter case, the extrapolation might produce undesired outliers. In our experiments, the datasets we used (SYMSOL, SYMSOL-T, and T-LESS) has abundant data that can cover the whole SO(3) and SE(3). We didn't do the experiment, but it would be interesting to investigate the sample efficiency of our score-based generative model.
2. One problem with using the 6D representation is that it requires the Gram-Schmidt process to orthogonalize the rotation matrix after the diffusion process. As we tested before, this might cause performance degradation similar to the quaternion case. On the other hand, our method, which predicts the score vector in Lie algebra and then diffuses on the Lie group, does not require normalization.

Feel free to open the issue if this does not address your question.

**Comment by bmeCEA** [comment]
Dear authors,

thank you for the great work!
I would like to elaborate on @sseunghyuns question:
I was wondering if you would be willing to share and explain the part of the code that iterates over the diffusion network and generate results such as these of figure 4 in the CVPR paper?

Thank you.

**Comment by Ending2015a** [comment]
Hello @bmeCEA ,
The visualization code is [here](https://github.com/Ending2015a/liepose-diffusion/blob/main/src/liepose/utils/vis.py#L31-L184).
In Figure 4, we perform inference with 1000 random poses on one image and use the above code to plot the results. You can edit the exp code [here](https://github.com/Ending2015a/liepose-diffusion/blob/main/src/liepose/exp/testbed.py#L448) to call the visualization function.
